
<html>
<head></head>
<body>

<style type="text/css">
body {
font-size:10pt;
}
</style>
<div style="margin:auto;width:650px;">

<div>

<h1>Genome-wide survey of sitewise selective pressures in mammals</h1>
<p style="text-align:right;">
The <a href="http://www.ensembl.org/">EnsEMBL Compara</a> team and <a href="http://www.ebi.ac.uk/goldman/">Goldman Group</a>, EMBL-EBI, Hinxton UK</a>
<br/>
Gregory Jordan, Albert Vilella, Tim Massingham, Ewan Birney and Nick Goldman
</p>
The datasets and UCSC genome browser tracks below contain the results from calculating the site-wise selective pressures acting on protein-coding regions of
all available vertebrate genomes, including the 22 mammals sequenced as a part of the Mammalian Genome Project.
</p>
<p>
Evolutionary pressures in four subtrees were considered in this analysis: vertebrates, primates, glires, and laurasiatheria. All of the datasets and many of the browser tracks
below contain separate tracks from independent calculations on each subtree, allowing for direct comparison of evolutionary pressures between 
distinct mammalian subtrees at various scales.
</p>
<p>
We used the protein-based gene tree alignments from the Ensembl Compara database [1] and the Sitewise Likelihood Ratio method
[2] to calculate dN/dS estimates at every protein-coding codon. Various filtering steps were employed in order to minimize errors resulting from
alignment, sequencing and annotation error.
</p>
<p>
References:<br/>
&nbsp;&nbsp; [1] Vilella, A. J. J. et al.  <a href="http://dx.doi.org/10.1101/gr.073585.107">EnsemblCompara GeneTrees: Analysis of complete, duplication aware phylogenetic trees in vertebrates.</a> Genome research (2008).
<br/>
&nbsp;&nbsp; [2] Massingham, T. & Goldman, N. <a href="http://dx.doi.org/10.1534/genetics.104.032144">Detecting amino acid sites under positive selection and purifying selection.</a> Genetics 169, 1753-1762 (2005).
</p>

</div>

<div style="border:2px solid #DDD;padding:1em;">
<h2>Contents</h2>

<h3>Gene summaries</h3>
<ul>
<li><b>By chromosome:</b> CHROMOSOME_LINKS </li>
<li><b>Top 200 by:</b> TOP_LINKS </li>
<li><b>Bottom 200 by:</b> BOTTOM_LINKS </li>
</ul>

<h3>UCSC Tracks</h3>
<ul>
<li><b>Sitewise dN/dS:</b> SITEWISE_LINKS </li>
<li><b>Genes colored by:</b> GENE_LINKS </li>
</ul>

<h3>Raw Data</h3>

</div>


</div>
</body>
</html>
